### PR TITLE
KAFKA-14114: Adding Metadata Log Processing Error Related Metrics

### PR DESCRIPTION
**Pull request by niket-goel**
This PR adds in 3 metrics as described in [KIP-859](https://cwiki.apache.org/confluence/display/KAFKA/KIP-859%3A+Add+Metadata+Log+Processing+Error+Related+Metrics)
* kafka.server:type=broker-metadata-metrics,name=metadata-load-error-count
* kafka.server:type=broker-metadata-metrics,name=metadata-apply-error-count
* kafka.controller:type=KafkaController,name=MetadataErrorCount

### CODE REVIEW

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 1]
```diff
@@ -114,6 +114,11 @@ class BrokerMetadataListener(
           debug(s"Loaded new commits: $loadResults")
         }
         loadResults
+      } catch {
+        case e: Throwable =>
+          brokerMetrics.listenerBatchLoadErrorCount.getAndIncrement()
+          log.error(e.toString)
+          BatchLoadResults(0, 0, 0, 0)
```
Comment (niket-goel) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: This is the most sketchy part of the change. I was thinking of what a good way to handle errors in this should be. Is there something that would need to be undone in this scenario?

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 2]
```diff
@@ -169,6 +174,10 @@ class BrokerMetadataListener(
         _delta.finishSnapshot()
         info(s"Loaded snapshot ${reader.snapshotId().offset}-${reader.snapshotId().epoch}: " +
           s"$loadResults")
+      } catch {
+        case e: Throwable =>
+          brokerMetrics.listenerBatchLoadErrorCount.getAndIncrement()
+          log.error(e.toString)
```
Comment (niket-goel) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: ditto

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerMetricsTest.java [hunk 3]
```diff
@@ -151,10 +154,18 @@ private static MetricName metricName(String type, String name) {
     }
 
     private static void assertMetricsCreated(MetricsRegistry registry, Set<String> expectedMetricNames, String expectedType) {
+        assertEquals(registry.allMetrics().keySet().stream()
```
Comment (niket-goel) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerMetricsTest.java:176]: This test was not testing if any metrics other than the expected existed.

File: metadata/src/main/java/org/apache/kafka/controller/QuorumControllerMetrics.java [hunk 4]
```diff
@@ -276,12 +297,15 @@ public long lastAppliedRecordTimestamp() {
     public void close() {
         Arrays.asList(
             ACTIVE_CONTROLLER_COUNT,
+            FENCED_BROKER_COUNT,
```
Comment (niket-goel) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumControllerMetrics.java:300]: Fixing the corresponding test surfaced the fact that we were not removing these metrics from the registry. Not sure if that was on purpose.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Why do we want to change the error handling here? Why don't we just rethrow the exception?

Comment (niket-goel) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: From what I understand, exceptions thrown here are handled by the super class `EventQueue.FailureLoggingEvent` which logs the error and moves on. I was reviewing if we need some special handling for the exception here in terms of possibly rolling back changes to any objects that were altered using this batch before we continue on.

Comment (hachikuji) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: That does not seem in the scope for a metrics change. Would it make sense to add the metric update to `FailureLoggingEvent` instead?

Comment (niket-goel) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala]: Had a conversation with @hachikuji offline. We can review the error handling as a separate PR. This does not seem to be the right place to alter it.

**Comment by cmccabe** [comment]
Thanks for the PR, @niket-goel. Looks like this needs to be rebased on trunk.

This needs to be integrated with the fault handler work. Specifically, the fault handlers need to increment the metrics, rather than doing it directly. This will allow juinit tests to monitor the fault handlers and make sure they are not triggering.

I was going to make a PR against your PR, but it ended up being too much work. The big hassle is the need to pull the metrics objects (BrokerMetrics, ControllerMetrics) up into KafkaRaftServer. Not conceptually difficult but a lot of churn.

This is the general idea, though:

```
   import BrokerMetadataPublisher._
@@ -259,9 +262,8 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
       }
       publishedOffsetAtomic.set(newImage.highestOffsetAndEpoch().offset)
     } catch {
-      case t: Throwable => error(s"Error publishing broker metadata at $highestOffsetAndEpoch", t)
-        brokerMetrics.metadataApplyErrorCount.getAndIncrement()
-        throw t
+      case t: Throwable =>
+        metadataPublishingFaultHandler.handleFault("Error publishing broker metadata at $highestOffsetAndEpoch", t)
     } finally {
       _firstPublish = false
     }
```

```
diff --git a/metadata/src/main/java/org/apache/kafka/metadata/fault/MetadataFaultHandler.java b/metadata/src/main/java/org/apache/kafka/metadata/fault/MetadataFaultHandler.java
index e9f71b80e6..7ccade98f9 100644
--- a/metadata/src/main/java/org/apache/kafka/metadata/fault/MetadataFaultHandler.java
+++ b/metadata/src/main/java/org/apache/kafka/metadata/fault/MetadataFaultHandler.java
@@ -27,10 +27,26 @@ import org.slf4j.LoggerFactory;
  */
 public class MetadataFaultHandler implements FaultHandler {
     private static final Logger log = LoggerFactory.getLogger(MetadataFaultHandler.class);
+    private final String prefix;
+    private final Runnable action;
+
+    public MetadataFaultHandler(
+        String prefix,
+        Runnable action
+    ) {
+        this.prefix = prefix;
+        this.action = action;
+    }
 
     @Override
     public void handleFault(String failureMessage, Throwable cause) {
+        failureMessage = prefix + ": " + failureMessage;
         FaultHandler.logFailureMessage(log, failureMessage, cause);
+        try {
+            action.run();
+        } catch (Throwable e) {
+            log.error("Failed to run MetadataFaultHandler action.", e);
+        }
         throw new MetadataFaultException("Encountered metadata fault: " + failureMessage, cause);
     }
 }
```

```
diff --git a/core/src/main/scala/kafka/server/KafkaRaftServer.scala b/core/src/main/scala/kafka/server/KafkaRaftServer.scala
index e7cf8f8f1f..bf3f5e828e 100644
--- a/core/src/main/scala/kafka/server/KafkaRaftServer.scala
+++ b/core/src/main/scala/kafka/server/KafkaRaftServer.scala
@@ -83,6 +83,8 @@ class KafkaRaftServer(
   )
 
   private val broker: Option[BrokerServer] = if (config.processRoles.contains(BrokerRole)) {
+
+
     Some(new BrokerServer(
       config,
       metaProps,
@@ -91,7 +93,11 @@ class KafkaRaftServer(
       metrics,
       threadNamePrefix,
       offlineDirs,
-      controllerQuorumVotersFuture
+      controllerQuorumVotersFuture,
+      new MetadataFaultHandler("Unable to load metadata",
+        () => brokerMetrics.metadataLoadErrorCount.getAndIncrement()),
+    new MetadataFaultHandler("Unable to publish metadata",
+        () => brokerMetrics.metadataApplyErrorCount.getAndIncrement())
     ))
   } else {
```

We also need to integrate this into the test harness so that if the fault handler is called, the test is failed.

This is how we did it in `QuorumTestHarness.scala`:
```
  @AfterEach
  def tearDown(): Unit = {
    Exit.resetExitProcedure()
    Exit.resetHaltProcedure()
    if (implementation != null) {
      implementation.shutdown()
    }
    System.clearProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM)
    Configuration.setConfiguration(null)
    faultHandler.maybeRethrowFirstException() <==== HERE
  }
``

and in `KafkaClusterTestKit.java`:
```
    @Override
    public void close() throws Exception {
...
        metadataFaultHandler.maybeRethrowFirstException();
        fatalFaultHandler.maybeRethrowFirstException();
    }
```

In these junit tests, it's probably OK to use the same `metadataFaultHandler` object for all the brokers, by default. We don't expect ANY broker to have a metadata fault, after all. If a test wants to do something exotic we can accommodate that in a custom way.

**Comment by niket-goel** [comment]
Thanks for the comments @cmccabe. I have rebased and integrated the metrics with the new Fault handling framework you added. Could you please take another look.

**Comment by niket-goel** [comment]
https://github.com/apache/kafka/pull/12496 Contains all changes in this PR and some more additional improvements. We will continue with that PR for this Improvement.

**Comment by cmccabe** [comment]
Merged from command line with the changes we discussed. Thanks, @niket-goel

**Pull request closed without merging**